design/20220720-per-certificate-owner-ref.md

```diff
@@ -0,0 +1,259 @@
+# Design: Per-Certificate Secret Owner Reference
+
+> 🌟 This design document was written by Maël Valais on 20 July 2022 in order to facilitate Denis Romanenko's feature request presented in [#5158](https://github.com/cert-manager/cert-manager/pull/5158).
+
+- [Release Signoff Checklist](#release-signoff-checklist)
+- [Summary](#summary)
+- [Use-cases](#use-cases)
+- [Questions](#questions)
+- [Proposal](#proposal)
+- [Design Details](#design-details)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Supported Versions](#supported-versions)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+This checklist contains actions which must be completed before a PR implementing this design can be merged.
+
+- [ ] This design doc has been discussed and approved
+- [ ] Test plan has been agreed upon and the tests implemented
+- [ ] Feature gate status has been agreed upon (whether the new functionality will be placed behind a feature gate or not)
+- [ ] Graduation criteria is in place if required (if the new functionality is placed behind a feature gate, how will it graduate between stages)
+- [ ] User-facing documentation has been PR-ed against the release branch in [cert-manager/website](https://github.com/cert-manager/website)
+
+## Summary
+
+The flag `--enable-certificate-owner-ref` allows you to configure cert-manager to delete Secret resources when the associated Certificate is removed. 
+
+We propose to introduce the same setting at the Certificate level so that users of the Certificate resource can decide whether the Secret resource should be removed or not.
+
+## Use-cases
+
+**Use-case 1: managed cert-manager installations**
+
+[Flant](https://flant.com) manages Kubernetes clusters for their customers. The installation of cert-manager is managed by Flant. Flant uses `--enable-certificate-owner-ref=false` to lower the chance of outages of their managed components. On the other hand, customers are relying on long-lived “dev” namespaces in which they install and uninstall their applications over and over with random names. The Certificate resources are correctly removed, but the Secret resources stay and accumulate.
+
+Source: https://github.com/deckhouse/deckhouse/pull/1601
+
```

Comment:
I object to this proposal because it does not solve the stated problem and because there are ways to solve this problem more generally, which do not require changes to cert-manager.

### Alternatives

The problem described in this use-case is that the "customers" are creating too many resources in their allocated namespace; so many that they are overloading the Flant Kubernetes API server.

One good answer to that problem is to impose an [Object Count Resource Quota](https://kubernetes.io/docs/concepts/policy/resource-quotas/#object-count-quota) which presumably would cause cert-manager to attach a meaningful "quota exceeded" error message to the `Certificate` resource. 
The developer would then know that they had to clean up stale resources from their dev namespace.

If I were such a developer I would attach a label to all the resources that I create during a given deployment,
and then delete all those resources using a label selector (`kubectl delete all --namespace=dev --label-selector=deployment-test-id=*` ).

Or if I were such a developer I would like to able to recreate a pristine dev namespace, perhaps by adding a label to my namespace. (e.g. `kubectl label ns dev recreate=daily`)

Or Flant could automatically delete any resources in "dev" namespaces which are older than X days.

Or they could implement an admission controller which forced developers to label all resources with a `time-to-live` label, and then delete any which outlived that duration.

My point is that there are many ways to solve this problem generally.

### Adding a `Certificate.Spec.CleanupPolicy` field does not **alone** solve the problem.

How will a new developer with a newly created "dev" namespace, know that it is their orphaned Secret resources that are causing the platform to become slow / unreliable?

How will that developer learn about this new field in the cert-manager Certificate resource?

And will they realise that it should be set to `cleanupPolicy: Never` in production and `cleanupPolicy: Always` in their "dev" namespaces?

If there were a way to add this field automatically to Certificate resources in certain namespaces, then I might be more convinced, but it still would not solve the general problem of "too many resources slowing down the K8S API server"

### Adding a `Certificate.Spec.CleanupPolicy` field does not solve **general** problem.

What happens when developer decides to use another operator that creates un-owned resources?
Won't they then have to find a general solution to this problem?

What happens if a developer writes a shell script that creates too many Secrets?

Reply:
@wallrj hello, you are right, it is many possible ways to solve problem of orphaned certs. But cert-manager alreeady have general way to do it - enable certificate owner reference. But it is not flexible - you cannot manage this per-certificate. We solve this problem by enabling global flag --certificate-owner-ref=true and disable owner reference in certificate CR for system certificates, so developer secrets removed after removing appropriate certificates, and system certificate secrets never removed. In another cases can be handy to disable global certificate owner ref parameter and enable it on some certificates. My point is that if it is possible to set the owner of the certificate ref, it should be flexible.

Reply:
I see your point but my (controversial) view is that the `--certificate-owner-ref` feature was a mistake,
and I think it should be deprecated and removed....**not** promoted and made more flexible.

I am strongly in favour of documenting how users can find Secrets that:
1. Have labels suggesting that they are (or were) managed by cert-manager.
2. No longer have a corresponding Certificate resource.
3. Are not mounted in any Pod.
4. Are not referenced by any Ingress.

I am in favour of a `cmctl experimental find-orphaned-secrets` command (or something similar) that performs some or all of those steps.

Reply:
I understand, yes, it's a different view of the problem. We can close this PR and deprecate `--certificate-owner-ref` option.